### PR TITLE
Add a configuration option and calculate the SubState Accumulator Hash and Epoch Hash

### PR DESCRIPTION
**Pull request by diegomrsantos**

### CODE REVIEW

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java [hunk 1]
```diff
@@ -296,5 +297,11 @@ protected void configure() {
     var enableTransactions = properties.get("api.transactions.enable", false);
     var enableSign = properties.get("api.sign.enable", false);
     install(new ApiModule(bindAddress, port, enableTransactions, enableSign));
+
+    // SubState Hash Accumulator
```
Comment (LukasGasior1) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java:301]: perhaps we should stick to `Substate` (a single word)?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.Optional;
+import java.util.function.Function;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bouncycastle.util.Arrays;
+
+public class BerkeleySubStateAccumulatorHashStore implements BerkeleyAdditionalStore {
+
+  private static final Logger logger = LogManager.getLogger();
+  protected static final byte[] SUBSTATE_ACCUMULATOR_HASH =
```
Comment (LukasGasior1) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:90]: SUBSTATE_ACCUMULATOR_HASH_KEY perhaps?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  private Optional<byte[]> getPreviousSubStateAccumulatorHash(Transaction dbTxn) {
+    var previousSubStateAccumulatorHash = new DatabaseEntry();
+    var key = new DatabaseEntry(SUBSTATE_ACCUMULATOR_HASH);
+    this.subStateAccumulatorHashDatabase.get(dbTxn, key, previousSubStateAccumulatorHash, null);
+    return Optional.ofNullable(previousSubStateAccumulatorHash.getData());
+  }
+
+  private long getEpoch() {
```
Comment (LukasGasior1) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:189]: how about something like: `getLatestStoredEpoch()`

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  private byte[] getBytes(REStateUpdate reStateUpdate) {
+    var op = reStateUpdate.isBootUp() ? new byte[] {0} : new byte[] {1};
+    var id = reStateUpdate.getId() != null ? reStateUpdate.getId().asBytes() : new byte[0];
+    var type = new byte[] {reStateUpdate.typeByte()};
+    var parsed = getParticleBytes(getParticle(reStateUpdate.getParsed()));
+    var stateBuf =
+        reStateUpdate.getStateBuf() != null
+            ? reStateUpdate.getRawSubstateBytes().getData()
```
Comment (LukasGasior1) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:205]: It seems that `getRawSubstateBytes` is a side-effecting operation: it consumes the bytes from the `stateBuf`. I wonder if there are any unobvious implications of this.

Comment (LukasGasior1) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:205]: hmh, or is this covered by the use of `Supplier` there :thinking:

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java (hunk 1 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java:301]: IntelliJ complains about that, but seems we use it like you said throughout the code, right?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java [hunk 5]
```diff
@@ -296,5 +297,11 @@ protected void configure() {
...
+    if (properties.get("substate.hash.accumulator.enable", false)) {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java:302]: I just followed the pattern used in `RadixNodeModule` and `default.config `files.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  private Particle getParticle(Object parsed) {
+    if (!(parsed instanceof Particle)) {
+      throw new IllegalStateException("REStateUpdate parsed field should be a Particle");
+    }
+    return (Particle) parsed;
+  }
+
+  private byte[] getParticleBytes(Particle particle) {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:218]: If we are going to rename it soon, then I agree. But I believe we should avoid this kind of approach in the future and either rename everything to a new name or stick with the old one throughout the code.

[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:218]: How about just `getBytes(Particle particle)`?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    var stateBuf =
+        reStateUpdate.getStateBuf() != null
+            ? reStateUpdate.getRawSubstateBytes().getData()
+            : new byte[0];
+    var instructionIndex = new byte[] {(byte) reStateUpdate.getInstructionIndex()};
+    return Arrays.concatenate(Arrays.concatenate(op, id, type, parsed), stateBuf, instructionIndex);
+  }
+
+  private Particle getParticle(Object parsed) {
+    if (!(parsed instanceof Particle)) {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:212]: I don't know. could we @talekhinezh?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 4 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:205]: I think you are right, not sure Supplier makes any difference here. Actually, I don't know if we need to use this field, @talekhinezh ?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java (hunk 1 above)
Comment (LukasGasior1) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java:301]: yes

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java:302]: I think we follow this pattern in most places, for example: `network.p2p.max_inbound_channels` or `db.cache_size.max`

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:218]: sounds good, or just `serialize(...)`

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:212]: I think that'll work

[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java:212]: But now sure if it is worth it to do it now, maybe it is better to do that as part of the refactoring.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java (hunk 1 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java]: done

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/RadixNodeModule.java]: done

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 2 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java]: done

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+  private Database subStateAccumulatorHashDatabase;
+  private Database epochHashDatabase;
+
+  private byte[] currentSubStateAccumulatorHash;
+  private long epoch;
+
+  private final Stopwatch watch = Stopwatch.createUnstarted();
+
+  @Inject private Forks forks;
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java]: done

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 3 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java]: done

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 4 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java]: @LukasGasior1 actually, it seems you are right about `Supplier` and a new `ByteBuffer` is returned every time. Please take a look at the new code.

File: radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REStateUpdate.java [hunk 9]
```diff
@@ -137,10 +137,15 @@ public Object getParsed() {
   }
 
   public RawSubstateBytes getRawSubstateBytes() {
```
Comment (diegomrsantos) [inline radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REStateUpdate.java:139]: @LukasGasior1 @siy does this change make sense?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubStateAccumulatorHashStore.java]: I changed it, seems to be safe.

**Pull request merged**